lib/logstash/inputs/kafka.rb

```diff
@@ -133,31 +134,32 @@ def run(logstash_queue)
     @logger.info('Running kafka', :group_id => @group_id, :topic_id => @topic_id, :zk_connect => @zk_connect)
     begin
       @consumer_group.run(@consumer_threads,@kafka_client_queue)
-      begin
-        while true
+
+      while !stop?
+        if !@kafka_client_queue.empty?
           event = @kafka_client_queue.pop
           queue_event(event, logstash_queue)
         end
-      rescue LogStash::ShutdownSignal
-        @logger.info('Kafka got shutdown signal')
-        @consumer_group.shutdown
       end
+
       until @kafka_client_queue.empty?
         queue_event(@kafka_client_queue.pop,logstash_queue)
       end
+
       @logger.info('Done running kafka input')
     rescue => e
       @logger.warn('kafka client threw exception, restarting',
                    :exception => e)
-      if @consumer_group.running?
-        @consumer_group.shutdown
-      end
-      sleep(Float(@consumer_restart_sleep_ms) * 1 / 1000)
-      retry
+      Stud.stoppable_sleep(Float(@consumer_restart_sleep_ms) * 1 / 1000) { stop? }
```

Comment:
iono, I didn't want to edit this. looks like `@consumer_restart_sleep_ms` is just a `number` so this forces it to be a float for more precision.